util/sokol_spritebatch.h

```diff
@@ -0,0 +1,1263 @@
+
+    _sbatch_vertex* top_right_vertex = vertices + 1;
+    top_right_vertex->x = x + w;
+    top_right_vertex->y = y;
+    top_right_vertex->z = depth;
+    top_right_vertex->rgba = rgba;
+    top_right_vertex->u = bottom_right.x;
+    top_right_vertex->v = top_left.y;
+
+    _sbatch_vertex* bottom_left_vertex = vertices + 2;
+    bottom_left_vertex->x = x;
+    bottom_left_vertex->y = y + h;
+    bottom_left_vertex->z = depth;
+    bottom_left_vertex->rgba = rgba;
+    bottom_left_vertex->u = top_left.x;
+    bottom_left_vertex->v = bottom_right.y;
+
+    _sbatch_vertex* bottom_right_vertex = vertices + 3;
+    bottom_right_vertex->x = x + w;
+    bottom_right_vertex->y = y + h;
+    bottom_right_vertex->z = depth;
+    bottom_right_vertex->rgba = rgba;
+    bottom_right_vertex->u = bottom_right.x;
+    bottom_right_vertex->v = bottom_right.y;
+}
+
+static void _sbatch_init_index_buffer(void) {
+    uint16_t* index_buffer = (uint16_t*)SOKOL_MALLOC(_SBATCH_MAX_INDICES * sizeof(uint16_t));
+    SOKOL_ASSERT(index_buffer);
+
+    uint16_t* index_ptr = index_buffer;
+    for (uint32_t i = 0; i < _SBATCH_MAX_QUADS; i++, index_ptr += 6) {
+        // Triangle 1
+        *(index_ptr + 0) = (uint16_t)(i * 4);
+        *(index_ptr + 1) = (uint16_t)(i * 4 + 1);
+        *(index_ptr + 2) = (uint16_t)(i * 4 + 2);
+        // Triangle 2
+        *(index_ptr + 3) = (uint16_t)(i * 4 + 1);
+        *(index_ptr + 4) = (uint16_t)(i * 4 + 3);
+        *(index_ptr + 5) = (uint16_t)(i * 4 + 2);
+    }
+
+    sg_buffer_desc index_buffer_desc;
+    memset(&index_buffer_desc, 0, sizeof(index_buffer_desc));
+    index_buffer_desc.size = _SBATCH_MAX_INDICES * sizeof(uint16_t);
+    index_buffer_desc.type = SG_BUFFERTYPE_INDEXBUFFER;
+    index_buffer_desc.usage = SG_USAGE_IMMUTABLE;
+    index_buffer_desc.label = "sokol-spritebatch-indices";
+    index_buffer_desc.data.size = _SBATCH_MAX_INDICES * sizeof(uint16_t);
+    index_buffer_desc.data.ptr = index_buffer;
+    _sbatch.index_buffer = sg_make_buffer(&index_buffer_desc);
+    SOKOL_ASSERT(SG_INVALID_ID != _sbatch.index_buffer.id);
+
+    SOKOL_FREE(index_buffer);
+
+    _sbatch.bindings.index_buffer = _sbatch.index_buffer;
+}
+
+static bool _sbatch_rect_is_valid(const sbatch_rect* rect) {
+    if (rect->width != 0.0f && rect->height != 0.0f) {
+        return true;
+    }
+    return false;
+}
+
+SOKOL_SPRITEBATCH_API_DECL int sbatch_frame() {
+    return ++_sbatch.frame_index;
+}
+
+SOKOL_API_IMPL void sbatch_setup(const sbatch_desc* desc) {
+
+    memset(&_sbatch, 0, sizeof _sbatch);
+
+    sbatch_desc batch_desc = *desc;
+    batch_desc.context_pool_size = _SBATCH_DEFAULT(batch_desc.context_pool_size, 32);
+    batch_desc.pipeline_pool_size = _SBATCH_DEFAULT(batch_desc.pipeline_pool_size, 32);
+    _sbatch_setup_context_pool(&batch_desc);
+    _sbatch_setup_pipeline_pool(&batch_desc);
+    _sbatch_setup_sprite_pool();
+
+    _sbatch.shader = sg_make_shader(spritebatch_shader_desc(sg_query_backend()));
```

Comment:
PS: if you can't run sokol-shdc on macOS (which just occured to me heh), just run sokol-shdc without the ```-b``` option, this generates source code for all platforms. I can then take care of replacing the generated shader source code one by one with bytecode (for D3D11, Metal macOS and Metal iOS).

Reply:
I ran this verbatim and it worked on windows.

 `sokol-shdc -i sgl.glsl -o sgl.h -l glsl330:glsl100:hlsl4:metal_macos:metal_ios:metal_sim:wgpu -b`

The output is what is now comitted to the source here. I don't expect it will have worked in the case of the apple platforms so if you have a good way of testing those then I'd really appreciate the assist.

Cheers!